src/Corax/Queries/BoostingMatch.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+
+namespace Corax.Queries
+{
+    public struct BoostingComparer : IMatchComparer
+    {
+        public MatchCompareFieldType FieldType => throw new NotImplementedException();
+
+        public int FieldId => throw new NotImplementedException();
+
+        public int CompareById(long idx, long idy)
+        {
+            throw new NotImplementedException();
+        }
+
+        public int CompareNumerical<T>(T sx, T sy) where T : unmanaged
+        {
+            throw new NotImplementedException();
+        }
+
+        public int CompareSequence(ReadOnlySpan<byte> sx, ReadOnlySpan<byte> sy)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    public unsafe struct BoostingMatch<TInner, TQueryScoreFunction> : IQueryMatch
+        where TInner : IQueryMatch
+        where TQueryScoreFunction : IQueryScoreFunction
+    {
+        internal readonly TInner _inner;
+        internal readonly TQueryScoreFunction _scorer;
+        private readonly delegate*<ref BoostingMatch<TInner, TQueryScoreFunction>, Span<long>, Span<float>, void> _scoreFunc;
+        
+        public BoostingMatch(in TInner inner, in TQueryScoreFunction scorer, delegate*<ref BoostingMatch<TInner, TQueryScoreFunction>, Span<long>, Span<float>, void> scoreFunc)
+        {
+            _inner = inner;
+            _scorer = scorer;
+            _scoreFunc = scoreFunc;
+        }
+
+        public long Count => _inner.Count;
+
+        public QueryCountConfidence Confidence => _inner.Confidence;
+
+        public bool IsBoosting => true;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public int AndWith(Span<long> prevMatches)
+        {
+            // TODO: Every time this is called we will store in a growable temporary buffer all the matches for this boosting.
```

Comment:
I don't understand this?

Reply:
The entire boosting logic resides inside the `BoostingMatch` with this design. Noone else have to know anything about boosting beyond been able to `.Score()` which in many cases it is just a `nop`

Reply:
So what is that about growable temporary buffer?

Reply:
The `BoostingMatch` has to know what matches came from this branch... you can have OR statements where some will be there and others will not be there. So we need to record every single match that passes through the `BoostingMatch` and then apply the multiplication only on those were there is a match. Because we dont know until it finishes how many matches there may be there (specially in the case of `MultiTermMatches`) we just reserve and grow that record over time until we are requested to 'score' the matches on `orderby` step (sorting).

Reply:
In that case, wouldn't it make more sense to combine the scoring with the `Fill` and `AndWith`?